### PR TITLE
[com_installer] [updatesites] Add options to delete and rebuild update sites

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/8512.
#### Summary of Changes

After updating Joomla and extensions a lot of times (since 1.7) we get a lot of obsolete update sites from domain changes, url changes, etc that extension developers forget to delete on uninstall.

This makes searching for updates more slow. Right now we can disable the update sites, but we cannot delete or even rebuild the update sites table.

This PR adds two new options to com_installer updates sites view: delete and rebuild
- delete: deletes an update site (joomla core updates sites cannot be deleted)
- rebuild: the update sites are rebuild from the manifest files.
#### Testing Instructions
1. Use latest staging
2. Apply this patch
3. Go to "Extensions -> Manage -> Update Sites"
4. Notice the two new buttons
   image
5. Try to Rebuild the update sites and check if everything is fine.
6. Install some extensions with update sites of all the types you can imagine (libraries, templates, plugin, modules, components, etc).
7. Repeat step 5 and check if everything is fine.
8. Delete some update sites (the 4 core joomla update sites will not be allowed to delete) the others you can delete.
9. Repeat step 5 and check if everything is fine.
#### Observations

Code revisions, improvements and suggestions are welcomed.

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 1]
```diff
@@ -131,6 +131,12 @@ COM_INSTALLER_MSG_UPDATE_NOUPDATES="There are no updates available at the moment
 COM_INSTALLER_MSG_UPDATE_SITES_COUNT_CHECK="Some update sites are disabled. You may want to check the <a href="_QQ_"%s"_QQ_">Update Sites Manager</a>."
 COM_INSTALLER_MSG_UPDATE_SUCCESS="Updating %s was successful."
 COM_INSTALLER_MSG_UPDATE_UPDATE="Update"
+COM_INSTALLER_MSG_UPDATESITES_DELETE_ERROR="And error as occured while trying to delete "_QQ_"%s"_QQ_" update site: %s."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]: An error has...

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 2]
```diff
@@ -131,6 +131,12 @@ COM_INSTALLER_MSG_UPDATE_NOUPDATES="There are no updates available at the moment
...
+COM_INSTALLER_MSG_UPDATESITES_DELETE_CANNOT_DELETE="%s update sites cannot be deleted."
+COM_INSTALLER_MSG_UPDATESITES_DELETE_UPDATESITES_DELETED="%s update sites have been deleted."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]: These two messages are plurals - do we also need a singular

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 3]
```diff
@@ -131,6 +131,12 @@ COM_INSTALLER_MSG_UPDATE_NOUPDATES="There are no updates available at the moment
...
+COM_INSTALLER_MSG_UPDATESITES_REBUILD_NOT_PERMITTED="Rebuild update sites is not permitted."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]: Rebuilding ...

File: administrator/language/en-GB/en-GB.com_installer.ini [hunk 4]
```diff
@@ -131,6 +131,12 @@ COM_INSTALLER_MSG_UPDATE_NOUPDATES="There are no updates available at the moment
...
+COM_INSTALLER_MSG_UPDATESITES_REBUILD_WARNING="Update sites have been rebuilt. No extension with updates sites discovered."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]: Not sure i this one is correct either - will need to find out exactly when it is displayed

File: administrator/language/en-GB/en-GB.com_installer.ini (hunk 2 above)
Comment (andrepereiradasilva) [inline administrator/language/en-GB/en-GB.com_installer.ini]: and how we do plural forms?

Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_installer.ini]: An example would be
COM_CONTENT_N_ITEMS_DELETED="%s articles deleted."
COM_CONTENT_N_ITEMS_DELETED_1="%s article deleted."

Comment (andrepereiradasilva) [inline administrator/language/en-GB/en-GB.com_installer.ini]: hum ok. i guess there's a method to use that in joomla api.

**Comment by andrepereiradasilva** [comment]
ok still need some changes here, plus resolving the plural forms

**Comment by infograf768** [comment]
@andrepereiradasilva 

I suggest to slightly change the constants for the plural strings, i.e adding _N_
from
```
+COM_INSTALLER_MSG_UPDATESITES_DELETE_UPDATESITES_DELETED="%s update sites have been deleted."
+COM_INSTALLER_MSG_UPDATESITES_DELETE_UPDATESITES_DELETED_1="1 update site has been deleted."
```
to
+COM_INSTALLER_MSG_UPDATESITES_DELETE_N_UPDATESITES_DELETED="%s update sites have been deleted."
+COM_INSTALLER_MSG_UPDATESITES_DELETE_N_UPDATESITES_DELETED_1="1 update site has been deleted."
```

**Comment by brianteeman** [comment]
Thanks @infograf768 that was what I had suggested before the _N_

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 17db21e4660002d9632e48dfd634c59e1e14b664<br /><br />Tested delete 
Tested rebuild
Added fake updates sites in the db and rebuild removed them

All good (once the language string key is updated)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9744">issues.joomla.org/joomla-cms/9744</a>.</sub>

**Comment by andrepereiradasilva** [comment]
please don't test yet i has some bugs still.
i'm trying to solve them.

**Comment by brianteeman** [comment]
Really? Every test I could think of worked

**Comment by andrepereiradasilva** [comment]
problems still to solve:
- the en-GB package, if somehow deleted, doesn't get the update (because it does not have an extension in the database ...). 
- also had to do an additional check if the extension is only discovered or it's really installed.

**Comment by brianteeman** [comment]
> the en-GB package, if somehow deleted, doesn't get the update (because it does not have an extension in the database ...).
 
Accredited Joomla! Translations update site cannot be deleted.

In addition its a protected extension so cant be deleted

**Comment by andrepereiradasilva** [comment]
@brianteeman 
After some investigation and with the help form @infograf768 the en-GB language problem is something not related to this PR.

Explaining, for what i know, and correct me if i'm wrong:

A language package install 3 extensions in the database:
- type = language | Admin
 - Manifest: /administrator/language/xx-XX/xx-XX.xml
- type = language | Site
 - Manifest: /language/xx-XX/xx-XX.xml
- type = package | Site
 - Manifest: /administrator/manifests/packages/pkg_xx-XX.xml

The en-GB language pack (comes by default) somehow only has two extensions in the database...
- type = language | Admin
 - Database Id: 601
 - Manifest: /administrator/language/en-GB/en-GB.xml
- type = language | Site
 - Database Id: 600
 - Manifest: /language/en-GB/en-GB.xml
- type = package | Site
 - Database Id: **doesn't exist**
 - Manifest: /administrator/manifests/packages/pkg_en-GB.xml

See https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L618-L619

Since the update servers are declared in the pkg manifest, if somehow deleted, will not be reconstructed in this PR.
Of course this is a wild scenario, since deleting it is not possible through the update sites view.

So this is not related to this PR, this PR does not readd the en-GB update site after "accidental" delete because, as said, the pkg_en-GB extension does not exist, by default, in joomla database.

But from a coherence point of view, if a manifest exists the extension should exist in the database.

In other words, IMHO it should exist a blocked en-GB package extension in the database on joomla install and the udpate site should point to that extension (as all language packs do).

Remember you have blocked extension for all other manifest files, including joomla core, joomla update, etc.

**Comment by brianteeman** [comment]
Not 100% sure I understand your last sentence but as for the rest it really is beyond the scope of this PR

**Comment by infograf768** [comment]
@brianteeman 
What @andrepereiradasilva  says is that this PR will work perfectly if we add in our sqls the Database Id: for the en-GB fake pack. 
If not done, this PR may let people delete the en-GB entry in updatesite by mistake. This would be an issue if someone has never installed a lang pack and does it for the first time after deleting it.
Just a matter of adding to joomla.sql and updates sql the fake pkg

**Comment by andrepereiradasilva** [comment]
yes, it's out of the scope of this PR.

> Remember you have blocked extension for all other manifest files, including joomla core, joomla update, etc.

Some examples: 

Joomla Core: All logic
- [Manifest](https://github.com/joomla/joomla-cms/blob/staging/administrator/manifests/files/joomla.xml#L51-L52)
- [Extension in db](https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L620)
- [Update sites in db](https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L1809)

Joomla Update: All logic
- [Manifest](https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_joomlaupdate/joomlaupdate.xml#L29)
- [Extension in db](https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L511)
- [Update sites in db](https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L1812)

en-GB language package: ?
- [Manifest](https://github.com/joomla/joomla-cms/blob/staging/administrator/manifests/packages/pkg_en-GB.xml#L22)
- Extension in db: **doesn't exist** 
- [Update sites in db](https://github.com/joomla/joomla-cms/blob/staging/installation/sql/mysql/joomla.sql#L1811)
 it's pointing to extension id 600 (en-GB site language) whose [manifest ](https://github.com/joomla/joomla-cms/blob/staging/language/en-GB/en-GB.xml) **doesn't have an updateserver** ...

Something doesn't seem right ...

Put ok, let's concentrate in the PR and check that later.

**Comment by andrepereiradasilva** [comment]
> What @andrepereiradasilva says is that this PR will work perfectly if we add in our sqls the Database Id: for the en-GB fake pack (in _extensions). 

Yes, but in another PR.

> If not done, this PR may let people delete the en-GB entry in updatesite by mistake. This would be an issue if someone has never installed a lang pack and does it for the first time after deleting it.

This PR doesn't allow to delete any update site that has `update.joomla.org` in the URI.
See https://github.com/andrepereiradasilva/joomla-cms/blob/update-sites-manage/administrator/components/com_installer/models/updatesites.php#L366

> Just a matter of adding to joomla.sql and updates sql the fake pkg.

Not only, the update site extension id should be linked en-Gb package too (since in the en-GB package that has the update server URI).

**Comment by brianteeman** [comment]
As this PR does not allow you to delete that update site I really think its a non-issue

**Comment by andrepereiradasilva** [comment]
Yes, it's an issue because of the special treatment that exists in en-GB, it's just an inconsistency i disocvered in this PR.

I will not fix this is this PR. So let's concentrate in this PR.

Will do the rest of the changes and tell when is finished.

**Comment by andrepereiradasilva** [comment]
ok, now should be ok.

**Comment by brianteeman** [comment]
I dont understand the comment about a discovered extension not being installed. If I click on discover and install an extension then it is installed and if it has an update server in the xml it is added to the update sites<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9744">issues.joomla.org/joomla-cms/9744</a>.</sub>

**Comment by brianteeman** [comment]
I realise now you meant not to parse the extensions that are listed in Discover but have not been installed yet<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9744">issues.joomla.org/joomla-cms/9744</a>.</sub>

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 4d26e5141300146fea41ed4ca9d000c905a43b30<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9744">issues.joomla.org/joomla-cms/9744</a>.</sub>

**Comment by conconnl** [comment]
I have tested this item :white_check_mark: successfully on 4d26e5141300146fea41ed4ca9d000c905a43b30<br /><br />I have tested it successfully.
Installed an extension, deleted the Update Site and recreated it by using the rebuild function.
Afterwards removed extension and the Update Site was removed too.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9744">issues.joomla.org/joomla-cms/9744</a>.</sub>

**Comment by brianteeman** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9744">issues.joomla.org/joomla-cms/9744</a>.</sub>

**Comment by insitevision** [comment]
I have tested this item :white_check_mark: successfully on 4d26e5141300146fea41ed4ca9d000c905a43b30<br /><br />Installed components, plugins, and template with update sites.
Followed the test instruction, deleted the update sites. After rebuild the update sites were succesfully recreated. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9744">issues.joomla.org/joomla-cms/9744</a>.</sub>
